src/Serilog.Sinks.Loggly/Sinks/Loggly/FileBufferDataProvider.cs

```diff
@@ -0,0 +1,279 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using Loggly;
+using Newtonsoft.Json;
+using Serilog.Debugging;
+
+namespace Serilog.Sinks.Loggly
+{
+    interface IBufferDataProvider
+    {
+        IEnumerable<LogglyEvent> GetBatchOfEvents();
+        void MarkCurrentBatchAsProcessed();
+        void MoveBookmarkFoward();
+    }
+
+    /// <summary>
+    /// Provides a facade to all File operations, namely bookmark management and 
+    /// buffered data readings
+    /// </summary>
+    class FileBufferDataProvider : IBufferDataProvider
+    {
+        readonly string _candidateSearchPath;
+        readonly string _logFolder;
+        readonly string _bookmarkFilename;
+
+        readonly int _batchPostingLimit;
+        readonly long? _eventBodyLimitBytes;
+
+        readonly IFileSystemAdapter _fileSystemAdapter;
+        readonly IBookmarkProvider _bookmarkProvider;
+        readonly Encoding _encoding;
+
+        readonly JsonSerializer _serializer = JsonSerializer.Create();
+
+        // the following fields control the internal state and position of the queue
+        Bookmark _currentBookmark;
+        Bookmark _futureBookmark;
+        IEnumerable<LogglyEvent> _currentBatchOfEventsToProcess;
+
+        public FileBufferDataProvider(
+            string baseBufferFileName, 
+            IFileSystemAdapter fileSystemAdapter, 
+            IBookmarkProvider bookmarkProvider,
+            Encoding encoding,
+            int batchPostingLimit,
+            long? eventBodyLimitBytes
+        )
+        {
+            _bookmarkFilename = Path.GetFullPath(baseBufferFileName + ".bookmark");
```

Comment:
this is the same file name as the one created by the FileBasedBookmarkProvider class. is this on purpose? seems redundant and error-prone

Reply:
yes, it's the same. There are two problems in that, really - FileBufferDataProvider doesn't need _bookmarkFileName, since it's no longer accessed directly by it. The only class that requires access to the bookmark file is FileBasedBookmarkProvider (so _bookmarkFulename can be removed from the buffer class).

On the other hand, the `bookmarkFilename` (constructed path) is being used to determine the folder in which the logs are being stored. `baseBufferFileName` should probably be 2 things, really - a log folder and filename prefix argument, or just assume a conventioned prefix ("buffer", for instance). It'll be less confusing, probably.